crates/biome_js_analyze/src/analyzers/suspicious/no_confusing_void_type.rs

```diff
@@ -53,63 +54,67 @@ declare_rule! {
     }
 }
 
-type Language = <AnyTsType as AstNode>::Language;
-
-// We only focus on union type
-pub enum VoidTypeIn {
+/// We only focus on union type
+pub enum VoidTypeContext {
     Union,
     Unknown,
 }
 
 impl Rule for NoConfusingVoidType {
-    type Query = Ast<AnyTsType>;
-    type State = VoidTypeIn;
+    type Query = Ast<TsVoidType>;
+    type State = VoidTypeContext;
     type Signals = Option<Self::State>;
     type Options = ();
 
     fn run(ctx: &RuleContext<Self>) -> Self::Signals {
         let node = ctx.query();
-
-        if let AnyTsType::TsVoidType(node) = node {
-            let result = node_in(node.syntax());
-            return result;
-        }
-
-        None
+        decide_void_type_context(node.syntax())
     }
+
     fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
         let node = ctx.query();
-        return Some(
-            RuleDiagnostic::new(
-                rule_category!(),
-                node.range(),
-                markup! {{match_message(state)}},
-            )
-            .note(markup! {
-                "Remove "<Emphasis>"void"</Emphasis>
-            }),
-        );
+        let message = match state {
+            VoidTypeContext::Union => "void is not valid as a constituent in a union type",
+            VoidTypeContext::Unknown => {
+                "void is only valid as a return type or a type argument in generic type"
+            }
+        };
+
+        Some(
+            RuleDiagnostic::new(rule_category!(), node.range(), markup! {{message}}).note(
+                markup! {
+                    "Remove "<Emphasis>"void"</Emphasis>
+                },
+            ),
+        )
     }
 }
 
-fn node_in(node: &SyntaxNode<Language>) -> Option<VoidTypeIn> {
+fn decide_void_type_context(node: &SyntaxNode<JsLanguage>) -> Option<VoidTypeContext> {
     for parent in node.parent()?.ancestors() {
         match parent.kind() {
+            JsSyntaxKind::TS_UNION_TYPE_VARIANT_LIST => {
+                // checks if the union type contains generic type
+                for children in parent.descendants() {
```

Comment:
I am not sure that descendants is the best choice here because it traverses the entire subtree. Instead, we should traverse every constituent and verify if the type is `void` or a parametrized type with `void` as argument.

This is the [approach used by the ESLint rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-invalid-void-type.ts#L148-L165).